File: old_csharp_src/ShadowEmpire/SimplePrefsWindowClass.cs
// Decompiled with JetBrains decompiler
// Type: WindowsApplication1.SimplePrefsWindowClass
// Assembly: WindowsApplication1, Version=1.0.8020.28903, Culture=neutral, PublicKeyToken=null
// MVID: F52869E5-0850-48AD-BBBE-68E7A4900AFE
// Assembly location: C:\Program Files (x86)\Steam\steamapps\common\Shadow Empire\ShadowEmpire.exe

using Microsoft.VisualBasic.CompilerServices;
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace WindowsApplication1
{
  public class SimplePrefsWindowClass : WindowClass
  {
    private int cancelid;
    private int Info1Id;
    private int info2id;
    private string ShowString;
    private DateTime ShowTime;
    private int w;
    private int h;
    private int CurrentView;
    private int BNameId;
    private int BNameTextId;
    private int B1Id;
    private int B1TextId;
    private int saveid;
    private int quitid;
    private int B2Id;
    private int B2TextId;
    private int B3Id;
    private int B3TextId;
    private int B4Id;
    private int B4TextId;
    private int B5Id;
    private int B5TextId;
    private int B6Id;
    private int B6TextId;
    private int B7Id;
    private int B7TextId;
    private int B8Id;
    private int B8TextId;
    private int B9Id;
    private int B9TextId;
    private int B10Id;
    private int B10TextId;
    private int B11Id;
    private int B11TextId;
    private int B12Id;
    private int B12TextId;
    private int B13Id;
    private int B13TextId;
    private int B14Id;
    private int B14TextId;
    private int B15Id;
    private int B15TextId;
    private int B16Id;
    private int B16TextId;
    private int slider1id;
    private int slider2id;

    public SimplePrefsWindowClass(ref GameClass tGame)
      : base(ref tGame, 480, 250, 8)
    {
      this.View();
    }

    public override void HandleToolTip(int x, int y)
    {
      base.HandleToolTip(x, y);
      if (this.SubPartCounter > -1)
      {
        int subPartCounter = this.SubPartCounter;
        for (int index = 0; index <= subPartCounter; ++index)
        {
          if (x > this.SubPartX[index] & x < this.SubPartX[index] + this.SubPartW[index] && y > this.SubPartY[index] & y < this.SubPartY[index] + this.SubPartH[index])
          {
            this.SubPartList[index].DescriptInfo(x - this.SubPartX[index], y - this.SubPartY[index]);
            if (Operators.CompareString(this.SubPartList[index].Descript, "", false) > 0)
            {
              this.game.EditObj.TipButton = true;
              this.game.EditObj.TipTitle = "";
              this.game.EditObj.TipText = this.SubPartList[index].Descript;
              return;
            }
          }
        }
      }
      int mouseCounter = this.MouseCounter;
      for (int index = 0; index <= mouseCounter; ++index)
      {
        if (x > this.MouseRect[index].X & x < this.MouseRect[index].X + this.MouseRect[index].Width && y > this.MouseRect[index].Y & y < this.MouseRect[index].Y + this.MouseRect[index].Height)
        {
          if (this.MouseData[index] > 0)
            this.game.EditObj.TipButton = true;
          this.game.EditObj.TipTitle = this.MouseTitle[index];
          this.game.EditObj.TipText = this.MouseText[index];
          break;
        }
      }
    }

    public void View()
    {
      if (this.cancelid > 0)
      {
        this.RemoveSubPart(this.cancelid);
        this.cancelid = 0;
      }
      if (this.B1Id > 0)
        this.RemoveSubPart(this.B1Id);
      if (this.B1TextId > 0)
        this.RemoveSubPart(this.B1TextId);
      if (this.B2Id > 0)
        this.RemoveSubPart(this.B2Id);
      if (this.B2TextId > 0)
        this.RemoveSubPart(this.B2TextId);
      if (this.B3Id > 0)
        this.RemoveSubPart(this.B3Id);
      if (this.B3TextId > 0)
        this.RemoveSubPart(this.B3TextId);
      if (this.B4Id > 0)
        this.RemoveSubPart(this.B4Id);
      if (this.B4TextId > 0)
        this.RemoveSubPart(this.B4TextId);
      if (this.B5Id > 0)
        this.RemoveSubPart(this.B5Id);
      if (this.B5TextId > 0)
        this.RemoveSubPart(this.B5TextId);
      if (this.B6Id > 0)
        this.RemoveSubPart(this.B6Id);
      if (this.B6TextId > 0)
        this.RemoveSubPart(this.B6TextId);
      if (this.B7Id > 0)
        this.RemoveSubPart(this.B7Id);
      if (this.B7TextId > 0)
        this.RemoveSubPart(this.B7TextId);
      if (this.B8Id > 0)
        this.RemoveSubPart(this.B8Id);
      if (this.B8TextId > 0)
        this.RemoveSubPart(this.B8TextId);
      if (this.B9Id > 0)
        this.RemoveSubPart(this.B9Id);
      if (this.B9TextId > 0)
        this.RemoveSubPart(this.B9TextId);
      if (this.B10Id > 0)
        this.RemoveSubPart(this.B10Id);
      if (this.B10TextId > 0)
        this.RemoveSubPart(this.B10TextId);
      if (this.B11Id > 0)
        this.RemoveSubPart(this.B11Id);
      if (this.B11TextId > 0)
        this.RemoveSubPart(this.B11TextId);
      if (this.B12Id > 0)
        this.RemoveSubPart(this.B12Id);
      if (this.B12TextId > 0)
        this.RemoveSubPart(this.B12TextId);
      if (this.B13Id > 0)
        this.RemoveSubPart(this.B13Id);
      if (this.B13TextId > 0)
        this.RemoveSubPart(this.B13TextId);
      if (this.B14Id > 0)
        this.RemoveSubPart(this.B14Id);
      if (this.B14TextId > 0)
        this.RemoveSubPart(this.B14TextId);
      if (this.B15Id > 0)
        this.RemoveSubPart(this.B15Id);
      if (this.B15TextId > 0)
        this.RemoveSubPart(this.B15TextId);
      if (this.B16Id > 0)
        this.RemoveSubPart(this.B16Id);
      if (this.B16TextId > 0)
        this.RemoveSubPart(this.B16TextId);
      if (this.slider1id > 0)
      {
        this.RemoveSubPart(this.slider1id);
        this.slider1id = 0;
      }
      if (this.slider2id > 0)
      {
        this.RemoveSubPart(this.slider2id);
        this.slider2id = 0;
      }
      this.ClearMouse();
      this.NewBackGroundAndClearAll(480, 250, -1);
      Graphics graphics = Graphics.FromImage((Image) this.OwnBitmap);
      graphics.SmoothingMode = SmoothingMode.HighQuality;
      graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;
      DrawMod.DrawMessFrame(ref this.OwnBitmap, ref graphics, 0, 0, 480, 250);
      this.BackBitmap = (Bitmap) this.OwnBitmap.Clone();
      if (!this.game.EditObj.SoundOn)
      {
        SubPartClass tsubpart = (SubPartClass) new MarcRadioPartClass(0, false, "Sound effects are currently turned off.", ref this.OwnBitmap, 40, 110);
        this.B1Id = this.AddSubPart(ref tsubpart, 40, 110, 35, 35, 1);
      }
      else
      {
        SubPartClass tsubpart = (SubPartClass) new MarcRadioPartClass(0, true, "Sound effects are  currently turned on.", ref this.OwnBitmap, 40, 110);
        this.B1Id = this.AddSubPart(ref tsubpart, 40, 110, 35, 35, 1);
      }
      DrawMod.DrawTextColouredMarc(ref graphics, "SOUND FX", this.game.MarcFont4, 90, 118, Color.White);
      if (!this.game.EditObj.IntroSoundOn)
      {
        SubPartClass tsubpart = (SubPartClass) new MarcRadioPartClass(0, false, "Music is currently turned off.", ref this.OwnBitmap, 40, 40);
        this.B8Id = this.AddSubPart(ref tsubpart, 40, 40, 35, 35, 1);
      }
      else
      {
        SubPartClass tsubpart = (SubPartClass) new MarcRadioPartClass(0, true, "Music is currently turned on.", ref this.OwnBitmap, 40, 40);
        this.B8Id = this.AddSubPart(ref tsubpart, 40, 40, 35, 35, 1);
      }
      DrawMod.DrawTextColouredMarc(ref graphics, "MUSIC", this.game.MarcFont4, 90, 48, Color.White);
      SubPartClass tsubpart1;
      if (this.slider1id < 1)
      {
        tsubpart1 = (SubPartClass) new NumberSliderSubPartClass2(this.game, "Music Volume = ", "%", 200, 0, 100, this.game.EditObj.Volume, tbackbitmap: (ref this.OwnBitmap), bbx: 230, bby: 30, tMarc: true);
        this.slider1id = this.AddSubPart(ref tsubpart1, 230, 30, 200, 40, 0);
      }
      if (this.slider2id < 1)
      {
        tsubpart1 = (SubPartClass) new NumberSliderSubPartClass2(this.game, "SFX Volume = ", "%", 200, 0, 100, this.game.EditObj.Volume2, tbackbitmap: (ref this.OwnBitmap), bbx: 230, bby: 100, tMarc: true);
        this.slider2id = this.AddSubPart(ref tsubpart1, 230, 100, 200, 40, 0);
      }
      tsubpart1 = (SubPartClass) new TextButtonPartClass("OK", 150, "Click to return to main screen. [ESC]", ref this.OwnBitmap, 165, 165, theight: 40, usefont: this.game.MarcFont4, useshadow: true, tMarcStyle: true);
      this.cancelid = this.AddSubPart(ref tsubpart1, 165, 165, 150, 40, 1);
    }

    public override WindowReturnClass HandleKeyPress(int nr, bool fromTimer = false)
    {
      WindowReturnClass windowReturnClass = new WindowReturnClass();
      try
      {
        if (nr == 27 | nr == 32)
        {
          windowReturnClass.AddCommand(6, 0);
          this.game.EditObj.Save(this.game.AppPath + "editobj.txt");
          windowReturnClass.SetFlag(true);
          return windowReturnClass;
        }
      }
      catch (Exception ex)
      {
        ProjectData.SetProjectError(ex);
        ProjectData.ClearProjectError();
      }
      return windowReturnClass;
    }

    public override WindowReturnClass HandleMouseClick(int x, int y, int b)
    {
      WindowReturnClass windowReturnClass = new WindowReturnClass();
      if (this.SubPartCounter > -1)
      {
        int subPartCounter = this.SubPartCounter;
        for (int index = 0; index <= subPartCounter; ++index)
        {
          if (x > this.SubPartX[index] & x < this.SubPartX[index] + this.SubPartW[index] && y > this.SubPartY[index] & y < this.SubPartY[index] + this.SubPartH[index])
          {
            int num = this.SubPartID[index];
            if (num == this.cancelid)
            {
              windowReturnClass.AddCommand(6, 0);
              this.game.EditObj.Save(this.game.AppPath + "editobj.txt");
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
            if (num == this.slider1id)
            {
              this.game.EditObj.Volume = this.SubPartList[index].Click(x - this.SubPartX[index], y - this.SubPartY[index], b);
              this.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              this.SubPartList[index].Scroller = true;
              SoundMod.ChangeEventSoundBg(this.game.EditObj);
              return windowReturnClass;
            }
            if (num == this.slider2id)
            {
              this.game.EditObj.Volume2 = this.SubPartList[index].Click(x - this.SubPartX[index], y - this.SubPartY[index], b);
              this.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              this.SubPartList[index].Scroller = true;
              SoundMod.ChangeEventSound(this.game.EditObj);
              return windowReturnClass;
            }
            if (num == this.B1Id)
            {
              this.game.EditObj.SoundOn = !this.game.EditObj.SoundOn;
              this.View();
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
            if (num == this.B8Id)
            {
              this.game.EditObj.IntroSoundOn = !this.game.EditObj.IntroSoundOn;
              if (this.game.EditObj.IntroSoundOn)
                SoundMod.PlayEventBackground(this.game.AppPath + "sound/" + this.game.ModOpeningSoundtrack, ref this.game.EditObj);
              if (!this.game.EditObj.IntroSoundOn)
                SoundMod.STopEventBackground();
              Application.DoEvents();
              this.View();
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
          }
        }
        windowReturnClass.SetFlag(false);
        return windowReturnClass;
      }
      windowReturnClass.SetFlag(false);
      return windowReturnClass;
    }

    public override WindowReturnClass HandleMouseUp(int x, int y, int b)
    {
      WindowReturnClass windowReturnClass = new WindowReturnClass();
      OrderResult orderResult = new OrderResult();
      if (this.SubPartCounter > -1)
      {
        int subPartCounter = this.SubPartCounter;
        for (int index = 0; index <= subPartCounter; ++index)
        {
          if (this.SubPartList[index].Scroller)
          {
            int num = this.SubPartID[index];
            if (num == this.slider1id)
            {
              this.game.EditObj.Volume = this.SubPartList[index].Click(x - this.SubPartX[index], y - this.SubPartY[index], b);
              this.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              this.SubPartList[index].Scroller = false;
              this.SubPartList[this.SubpartNr(this.slider2id)].Scroller = false;
              SoundMod.ChangeEventSoundBg(this.game.EditObj);
              return windowReturnClass;
            }
            if (num == this.slider2id)
            {
              this.game.EditObj.Volume2 = this.SubPartList[index].Click(x - this.SubPartX[index], y - this.SubPartY[index], b);
              this.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              this.SubPartList[index].Scroller = false;
              this.SubPartList[this.SubpartNr(this.slider1id)].Scroller = false;
              SoundMod.ChangeEventSound(this.game.EditObj);
              return windowReturnClass;
            }
          }
        }
        windowReturnClass.SetFlag(false);
        return windowReturnClass;
      }
      windowReturnClass.SetFlag(false);
      return windowReturnClass;
    }
  }
}

File: src/SimplePrefsWindowClass.rs
// Decompiled with JetBrains decompiler
// Type: WindowsApplication1.SimplePrefsWindowClass
// Assembly: WindowsApplication1, Version=1.0.8020.28903, Culture=neutral, PublicKeyToken=null
// MVID: F52869E5-0850-48AD-BBBE-68E7A4900AFE
// Assembly location: C:\Program Files (x86)\Steam\steamapps\common\Shadow Empire\ShadowEmpire.exe

// usingMicrosoft.VisualBasic.CompilerServices;
// usingSystem;
// usingSystem.Drawing;
// usingSystem.Drawing.Drawing2D;
// usingSystem.Windows.Forms;

namespace WindowsApplication1
{
  pub class SimplePrefsWindowClass : WindowClass
  {
     cancelid: i32;
     Info1Id: i32;
     info2id: i32;
     ShowString: String;
     DateTime ShowTime;
     w: i32;
     h: i32;
     CurrentView: i32;
     BNameId: i32;
     BNameTextId: i32;
     B1Id: i32;
     B1TextId: i32;
     saveid: i32;
     quitid: i32;
     B2Id: i32;
     B2TextId: i32;
     B3Id: i32;
     B3TextId: i32;
     B4Id: i32;
     B4TextId: i32;
     B5Id: i32;
     B5TextId: i32;
     B6Id: i32;
     B6TextId: i32;
     B7Id: i32;
     B7TextId: i32;
     B8Id: i32;
     B8TextId: i32;
     B9Id: i32;
     B9TextId: i32;
     B10Id: i32;
     B10TextId: i32;
     B11Id: i32;
     B11TextId: i32;
     B12Id: i32;
     B12TextId: i32;
     B13Id: i32;
     B13TextId: i32;
     B14Id: i32;
     B14TextId: i32;
     B15Id: i32;
     B15TextId: i32;
     B16Id: i32;
     B16TextId: i32;
     slider1id: i32;
     slider2id: i32;

    pub SimplePrefsWindowClass( tGame: GameClass)
      : base( tGame, 480, 250, 8)
    {
      self.View();
    }

    pub fn HandleToolTip(x: i32, y: i32)
    {
      base.HandleToolTip(x, y);
      if (self.SubPartCounter > -1)
      {
        let mut subPartCounter: i32 = self.SubPartCounter;
        for (let mut index: i32 = 0; index <= subPartCounter; index += 1)
        {
          if (x > self.SubPartX[index] & x < self.SubPartX[index] + self.SubPartW[index] && y > self.SubPartY[index] & y < self.SubPartY[index] + self.SubPartH[index])
          {
            self.SubPartList[index].DescriptInfo(x - self.SubPartX[index], y - self.SubPartY[index]);
            if (Operators.CompareString(self.SubPartList[index].Descript, "", false) > 0)
            {
              self.game.EditObj.TipButton = true;
              self.game.EditObj.TipTitle = "";
              self.game.EditObj.TipText = self.SubPartList[index].Descript;
              return;
            }
          }
        }
      }
      let mut mouseCounter: i32 = self.MouseCounter;
      for (let mut index: i32 = 0; index <= mouseCounter; index += 1)
      {
        if (x > self.MouseRect[index].X & x < self.MouseRect[index].X + self.MouseRect[index].Width && y > self.MouseRect[index].Y & y < self.MouseRect[index].Y + self.MouseRect[index].Height)
        {
          if (self.MouseData[index] > 0)
            self.game.EditObj.TipButton = true;
          self.game.EditObj.TipTitle = self.MouseTitle[index];
          self.game.EditObj.TipText = self.MouseText[index];
          break;
        }
      }
    }

    pub fn View()
    {
      if (self.cancelid > 0)
      {
        self.RemoveSubPart(self.cancelid);
        self.cancelid = 0;
      }
      if (self.B1Id > 0)
        self.RemoveSubPart(self.B1Id);
      if (self.B1TextId > 0)
        self.RemoveSubPart(self.B1TextId);
      if (self.B2Id > 0)
        self.RemoveSubPart(self.B2Id);
      if (self.B2TextId > 0)
        self.RemoveSubPart(self.B2TextId);
      if (self.B3Id > 0)
        self.RemoveSubPart(self.B3Id);
      if (self.B3TextId > 0)
        self.RemoveSubPart(self.B3TextId);
      if (self.B4Id > 0)
        self.RemoveSubPart(self.B4Id);
      if (self.B4TextId > 0)
        self.RemoveSubPart(self.B4TextId);
      if (self.B5Id > 0)
        self.RemoveSubPart(self.B5Id);
      if (self.B5TextId > 0)
        self.RemoveSubPart(self.B5TextId);
      if (self.B6Id > 0)
        self.RemoveSubPart(self.B6Id);
      if (self.B6TextId > 0)
        self.RemoveSubPart(self.B6TextId);
      if (self.B7Id > 0)
        self.RemoveSubPart(self.B7Id);
      if (self.B7TextId > 0)
        self.RemoveSubPart(self.B7TextId);
      if (self.B8Id > 0)
        self.RemoveSubPart(self.B8Id);
      if (self.B8TextId > 0)
        self.RemoveSubPart(self.B8TextId);
      if (self.B9Id > 0)
        self.RemoveSubPart(self.B9Id);
      if (self.B9TextId > 0)
        self.RemoveSubPart(self.B9TextId);
      if (self.B10Id > 0)
        self.RemoveSubPart(self.B10Id);
      if (self.B10TextId > 0)
        self.RemoveSubPart(self.B10TextId);
      if (self.B11Id > 0)
        self.RemoveSubPart(self.B11Id);
      if (self.B11TextId > 0)
        self.RemoveSubPart(self.B11TextId);
      if (self.B12Id > 0)
        self.RemoveSubPart(self.B12Id);
      if (self.B12TextId > 0)
        self.RemoveSubPart(self.B12TextId);
      if (self.B13Id > 0)
        self.RemoveSubPart(self.B13Id);
      if (self.B13TextId > 0)
        self.RemoveSubPart(self.B13TextId);
      if (self.B14Id > 0)
        self.RemoveSubPart(self.B14Id);
      if (self.B14TextId > 0)
        self.RemoveSubPart(self.B14TextId);
      if (self.B15Id > 0)
        self.RemoveSubPart(self.B15Id);
      if (self.B15TextId > 0)
        self.RemoveSubPart(self.B15TextId);
      if (self.B16Id > 0)
        self.RemoveSubPart(self.B16Id);
      if (self.B16TextId > 0)
        self.RemoveSubPart(self.B16TextId);
      if (self.slider1id > 0)
      {
        self.RemoveSubPart(self.slider1id);
        self.slider1id = 0;
      }
      if (self.slider2id > 0)
      {
        self.RemoveSubPart(self.slider2id);
        self.slider2id = 0;
      }
      self.ClearMouse();
      self.NewBackGroundAndClearAll(480, 250, -1);
      Graphics graphics = Graphics.FromImage((Image) self.OwnBitmap);
      graphics.SmoothingMode = SmoothingMode.HighQuality;
      graphics.InterpolationMode = InterpolationMode.HighQualityBicubic;
      DrawMod.DrawMessFrame( self.OwnBitmap,  graphics, 0, 0, 480, 250);
      self.BackBitmap = (Bitmap) self.OwnBitmap.Clone();
      if (!self.game.EditObj.SoundOn)
      {
        let mut tsubpart: SubPartClass =  new MarcRadioPartClass(0, false, "Sound effects are currently turned off.",  self.OwnBitmap, 40, 110);
        self.B1Id = self.AddSubPart( tsubpart, 40, 110, 35, 35, 1);
      }
      else
      {
        let mut tsubpart: SubPartClass =  new MarcRadioPartClass(0, true, "Sound effects are  currently turned on.",  self.OwnBitmap, 40, 110);
        self.B1Id = self.AddSubPart( tsubpart, 40, 110, 35, 35, 1);
      }
      DrawMod.DrawTextColouredMarc( graphics, "SOUND FX", self.game.MarcFont4, 90, 118, Color.White);
      if (!self.game.EditObj.IntroSoundOn)
      {
        let mut tsubpart: SubPartClass =  new MarcRadioPartClass(0, false, "Music is currently turned off.",  self.OwnBitmap, 40, 40);
        self.B8Id = self.AddSubPart( tsubpart, 40, 40, 35, 35, 1);
      }
      else
      {
        let mut tsubpart: SubPartClass =  new MarcRadioPartClass(0, true, "Music is currently turned on.",  self.OwnBitmap, 40, 40);
        self.B8Id = self.AddSubPart( tsubpart, 40, 40, 35, 35, 1);
      }
      DrawMod.DrawTextColouredMarc( graphics, "MUSIC", self.game.MarcFont4, 90, 48, Color.White);
      SubPartClass tsubpart1;
      if (self.slider1id < 1)
      {
        tsubpart1 =  new NumberSliderSubPartClass2(self.game, "Music Volume = ", "%", 200, 0, 100, self.game.EditObj.Volume, tbackbitmap: ( self.OwnBitmap), bbx: 230, bby: 30, tMarc: true);
        self.slider1id = self.AddSubPart( tsubpart1, 230, 30, 200, 40, 0);
      }
      if (self.slider2id < 1)
      {
        tsubpart1 =  new NumberSliderSubPartClass2(self.game, "SFX Volume = ", "%", 200, 0, 100, self.game.EditObj.Volume2, tbackbitmap: ( self.OwnBitmap), bbx: 230, bby: 100, tMarc: true);
        self.slider2id = self.AddSubPart( tsubpart1, 230, 100, 200, 40, 0);
      }
      tsubpart1 =  new TextButtonPartClass("OK", 150, "Click to return to main screen. [ESC]",  self.OwnBitmap, 165, 165, theight: 40, usefont: self.game.MarcFont4, useshadow: true, tMarcStyle: true);
      self.cancelid = self.AddSubPart( tsubpart1, 165, 165, 150, 40, 1);
    }

    pub HandleKeyPress: WindowReturnClass(nr: i32, bool fromTimer = false)
    {
      windowReturnClass: WindowReturnClass = WindowReturnClass::new();
      try
      {
        if (nr == 27 | nr == 32)
        {
          windowReturnClass.AddCommand(6, 0);
          self.game.EditObj.Save(self.game.AppPath + "editobj.txt");
          windowReturnClass.SetFlag(true);
          return windowReturnClass;
        }
      }
      catch (Exception ex)
      {
        ProjectData.SetProjectError(ex);
        ProjectData.ClearProjectError();
      }
      return windowReturnClass;
    }

    pub HandleMouseClick: WindowReturnClass(x: i32, y: i32, b: i32)
    {
      windowReturnClass: WindowReturnClass = WindowReturnClass::new();
      if (self.SubPartCounter > -1)
      {
        let mut subPartCounter: i32 = self.SubPartCounter;
        for (let mut index: i32 = 0; index <= subPartCounter; index += 1)
        {
          if (x > self.SubPartX[index] & x < self.SubPartX[index] + self.SubPartW[index] && y > self.SubPartY[index] & y < self.SubPartY[index] + self.SubPartH[index])
          {
            let mut num: i32 = self.SubPartID[index];
            if (num == self.cancelid)
            {
              windowReturnClass.AddCommand(6, 0);
              self.game.EditObj.Save(self.game.AppPath + "editobj.txt");
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
            if (num == self.slider1id)
            {
              self.game.EditObj.Volume = self.SubPartList[index].Click(x - self.SubPartX[index], y - self.SubPartY[index], b);
              self.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              self.SubPartList[index].Scroller = true;
              SoundMod.ChangeEventSoundBg(self.game.EditObj);
              return windowReturnClass;
            }
            if (num == self.slider2id)
            {
              self.game.EditObj.Volume2 = self.SubPartList[index].Click(x - self.SubPartX[index], y - self.SubPartY[index], b);
              self.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              self.SubPartList[index].Scroller = true;
              SoundMod.ChangeEventSound(self.game.EditObj);
              return windowReturnClass;
            }
            if (num == self.B1Id)
            {
              self.game.EditObj.SoundOn = !self.game.EditObj.SoundOn;
              self.View();
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
            if (num == self.B8Id)
            {
              self.game.EditObj.IntroSoundOn = !self.game.EditObj.IntroSoundOn;
              if (self.game.EditObj.IntroSoundOn)
                SoundMod.PlayEventBackground(self.game.AppPath + "sound/" + self.game.ModOpeningSoundtrack,  self.game.EditObj);
              if (!self.game.EditObj.IntroSoundOn)
                SoundMod.STopEventBackground();
              Application.DoEvents();
              self.View();
              windowReturnClass.SetFlag(true);
              return windowReturnClass;
            }
          }
        }
        windowReturnClass.SetFlag(false);
        return windowReturnClass;
      }
      windowReturnClass.SetFlag(false);
      return windowReturnClass;
    }

    pub HandleMouseUp: WindowReturnClass(x: i32, y: i32, b: i32)
    {
      windowReturnClass: WindowReturnClass = WindowReturnClass::new();
      OrderResult orderResult = OrderResult::new();
      if (self.SubPartCounter > -1)
      {
        let mut subPartCounter: i32 = self.SubPartCounter;
        for (let mut index: i32 = 0; index <= subPartCounter; index += 1)
        {
          if (self.SubPartList[index].Scroller)
          {
            let mut num: i32 = self.SubPartID[index];
            if (num == self.slider1id)
            {
              self.game.EditObj.Volume = self.SubPartList[index].Click(x - self.SubPartX[index], y - self.SubPartY[index], b);
              self.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              self.SubPartList[index].Scroller = false;
              self.SubPartList[self.SubpartNr(self.slider2id)].Scroller = false;
              SoundMod.ChangeEventSoundBg(self.game.EditObj);
              return windowReturnClass;
            }
            if (num == self.slider2id)
            {
              self.game.EditObj.Volume2 = self.SubPartList[index].Click(x - self.SubPartX[index], y - self.SubPartY[index], b);
              self.SubPartFlag[index] = true;
              windowReturnClass.SetFlag(true);
              self.SubPartList[index].Scroller = false;
              self.SubPartList[self.SubpartNr(self.slider1id)].Scroller = false;
              SoundMod.ChangeEventSound(self.game.EditObj);
              return windowReturnClass;
            }
          }
        }
        windowReturnClass.SetFlag(false);
        return windowReturnClass;
      }
      windowReturnClass.SetFlag(false);
      return windowReturnClass;
    }
  }
}
